webcrypto/hmac.go

```diff
@@ -185,3 +185,135 @@ func exportHMACKey(ck *CryptoKey, format KeyFormat) ([]byte, error) {
+// into `SubtleCrypto.GenerateKey`, when generating an HMAC key.
+type HmacImportParams struct {
+	Algorithm
+
+	// Hash represents the name of the digest function to use. You can
+	// use any of the following: [Sha256], [Sha384],
+	// or [Sha512].
+	Hash Algorithm `json:"hash"`
+
+	// Length holds (a Number) the length of the key, in bits.
+	// If this is omitted, the length of the key is equal to the block size
+	// of the hash function you have chosen.
+	// Unless you have a good reason to use a different length, omit
+	// use the default.
+	Length null.Int `json:"length"`
+}
+
+// newHmacImportParams creates a new HmacImportParams object from the given
+// algorithm and params objects.
+func newHmacImportParams(rt *goja.Runtime, normalized Algorithm, params goja.Value) (*HmacImportParams, error) {
+	// The specification doesn't explicitly tell us what to do if the
+	// hash field is not present, but we assume it's a mandatory field
+	// and throw an error if it's not present.
+	hashValue, err := traverseObject(rt, params, "hash")
+	if err != nil {
+		return nil, NewError(0, SyntaxError, "could not get hash from algorithm parameter")
+	}
+
+	// Although the specification doesn't explicitly ask us to do so, we
+	// normalize the hash algorithm here, as it shares the same definition
+	// as the AlgorithmIdentifier type, and we'll need it later on.
+	//
+	// Note @oleiade: The specification seems to assume that the normalization
+	// algorithm will normalize the existing Algorithm fields, and leave
+	// the rest untouched. As we are in the context of a statically typed
+	// language, we can't do that, so we need to normalize the hash
+	// algorithm here.
+	normalizedHash, err := normalizeAlgorithm(rt, hashValue, OperationIdentifierGenerateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// As the length attribute is optional and as the key generation process
+	// differentiates unset from zero-values, we need to handle the case
+	// where the length attribute is not present in the params object.
+	var length null.Int
+	algorithmLengthValue, err := traverseObject(rt, params, "length")
+	if err == nil {
+		length = null.IntFrom(algorithmLengthValue.ToInteger())
+	}
+
+	return &HmacImportParams{
+		Algorithm: normalized,
+		Hash:      normalizedHash,
+		Length:    length,
+	}, nil
+}
+
+// ImportKey imports a key from raw key data. It implements the KeyImporter interface.
+func (hip *HmacImportParams) ImportKey(
+	format KeyFormat,
+	keyData []byte,
+	keyUsages []CryptoKeyUsage,
+) (*CryptoKey, error) {
+	// 2.
+	for _, usage := range keyUsages {
+		switch usage {
+		case SignCryptoKeyUsage, VerifyCryptoKeyUsage:
+			continue
+		default:
+			return nil, NewError(0, SyntaxError, "invalid key usage: "+usage)
+		}
+	}
+
+	// 3.
+	var hash KeyAlgorithm
+
+	// 4.
+	switch format {
+	case RawKeyFormat:
+		hash = KeyAlgorithm{Algorithm{Name: hip.Hash.Name}}
+	default:
+		return nil, NewError(0, NotSupportedError, "unsupported key format "+format)
+	}
```

Comment:
Should we centralize this logic in the key algorithm?

Reply:
I'm not sure I understand what you mean, could you elaborate please? 😄

Reply:
I had the impression this was the same logic applied across different keys but it is not true. The switch is the same but the logic for the case is different. You can skip this.